src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -688,67 +918,187 @@ void ShenandoahFreeSet::rebuild() {
   log_debug(gc, free)("Rebuilding FreeSet");
   for (size_t idx = 0; idx < _heap->num_regions(); idx++) {
     ShenandoahHeapRegion* region = _heap->get_region(idx);
+
+    // We move all young available regions into mutator_free set and then we take back the regions we need for our
+    // reserve.  This allows us to "compact" the collector_free (survivor) regions at the high end of the heap.
+    clear_mutator_free(idx);
+    clear_collector_free(idx);
+    clear_old_collector_free(idx);
+
     if (region->is_alloc_allowed() || region->is_trash()) {
-      assert(!region->is_cset(), "Shouldn't be adding those to the free set");
+      assert(!region->is_cset(), "Shouldn't be adding cset regions to the free set");
 
       // Do not add regions that would surely fail allocation
       if (has_no_alloc_capacity(region)) continue;
 
-      _capacity += alloc_capacity(region);
-      assert(_used <= _capacity, "must not use more than we have");
-
-      assert(!is_mutator_free(idx), "We are about to add it, it shouldn't be there already");
-      _mutator_free_bitmap.set_bit(idx);
+      if (region->is_old()) {
+        _old_capacity += alloc_capacity(region);
+        assert(!is_old_collector_free(idx), "We are about to add it, it shouldn't be there already");
+        set_old_collector_free(idx);
+        log_debug(gc)("  Setting Region " SIZE_FORMAT " _old_collector_free_bitmap bit to true", idx);
+      } else {
+        _capacity += alloc_capacity(region);
 
-      log_debug(gc, free)("  Adding Region " SIZE_FORMAT " (Free: " SIZE_FORMAT "%s, Used: " SIZE_FORMAT "%s) to mutator free set",
+        assert(_used <= _capacity, "must not use more than we have");
+        assert(!is_mutator_free(idx), "We are about to add it, it shouldn't be there already");
+        set_mutator_free(idx);
+        log_debug(gc, free)(
+          "  Adding Region " SIZE_FORMAT " (Free: " SIZE_FORMAT "%s, Used: " SIZE_FORMAT "%s) to mutator free set",
           idx, byte_size_in_proper_unit(region->free()), proper_unit_for_byte_size(region->free()),
-               byte_size_in_proper_unit(region->used()), proper_unit_for_byte_size(region->used()));
+          byte_size_in_proper_unit(region->used()), proper_unit_for_byte_size(region->used()));
+      }
     }
   }
 
   // Evac reserve: reserve trailing space for evacuations
```

Comment:
I'd move the individual region assignment loop above into its own work method, and the remainder of the work below that adjusts the reserves into its own work method, if appropriate? Not sure if such a refactor makes sense or not, but might make the code arguable more maintainable/readable?

Reply:
I've tidied this code and added/improved comments to explain its behavior.  I've moved the first loop into its own work method.